### ISSUE TITLE
Shifting to CoffeeScript

### ISSUE THREAD

**Issue by ghost** [aneesh-neelam/VITacademics#31]
You guys wanna do a marathon shift of all the javascript to coffescript in a different branch or something will be happy to help you guys

**Comment by sreeram-dev** [comment]
What is the need?

**Comment by aneesh-neelam** [comment]
I don't think it is possible to shift the whole thing to CoffeeScript. Correct me if I'm wrong. I believe you can shift some part of it, but not the whole thing. Especially, when we're using so many custom JavaScript libraries like [async](https://github.com/caolan/async), [Underscore](http://underscorejs.org/) etc. Check out [package.json](https://github.com/aneesh-neelam/VITacademics/blob/master/package.json) and see if those libraries can be used with CoffeeScript as well. 

Also, it would be unnecessary and would make the entire codebase inconsistent and difficult to maintain, especially if we use a combination of both JavaScript and CoffeeScript. You are free to try, please do tell us how it goes. 

However, you must understand we do have priorities. If you can implement better, more optimized algorithms for scraping, storing and compiling the data pulled from VIT's server(s), that would be better. Also, we have a few features that we would like to be implemented, like semester grades, SMS service, a notice system, a WebApp and maybe Grunt/Gulp support. 

You can develop these new features in CoffeeScript if you can and if you like, but please, we would like the transformed and prettified JavaScript to be integarted with the rest of the JavaScript codebase. 

If you wish to contribute, please feel free to fork the repository and do so. Thank you for your suggestions, we'll take all the help we can get.

**Comment by ghost** [comment]
It will make the code much less verbose and the code will follow the strict indentation needs of coffescript which will make it much more readable which is important for an open source project like this especially if you want to bring in future collabrators

**Comment by aneesh-neelam** [comment]
We can use [JSLint](http://www.jslint.com/) for that particular purpose.

**Comment by ghost** [comment]
Coffescript compiles to javascript,irrespective of the libraries you use you can shift your entire js to coffee

**Comment by aneesh-neelam** [comment]
OK. If you believe this can be done, then you're free to try. But like I already mentioned, priorities. We're short on manpower.

**Comment by ghost** [comment]
How can you be short on manpower in VIT dude

**Comment by ghost** [comment]
And JSLint can never match coffescript that's for sure

**Comment by sreeram-dev** [comment]
@aashrairavooru make a repo. get a MVP ready. Aneesh has done a good job modularizing things. Codebase is easy to read. As your repo gains traction, contributions will come.
Your point is valid. But we do not have active contributors.

People who can contribute are busy doing their own projects. Majority are still learning to code.

**Comment by aneesh-neelam** [comment]
Obviously JSLint can't suit all the purposes of CoffeeScript. I didn't mean that. However, we believe it is unnecessarily a lot of effort, rewriting the whole codebase. 

However, I shall tell you something. I attempted to convince @saurabhsjoshi, @karthikb351, @biocross and @kishore-narendran to move to this Node.js Backend hosted on Heroku from their own Python GAE-hosted one only after it was almost complete. 

Please feel free to improve upon our existing codebase or rewrite it in a better fashion. We welcome any suggestions or improvements to this project.

**Comment by aneesh-neelam** [comment]
@aashrairavooru: Fork the codebase, make any improvements that you see fit. But do ensure that the integrity of the API Endpoints and the data is intact. 

We welcome contributions. However, I mentioned we're short on manpower because very few students are actually interested in contributing to other's projects. As @sreeram-boyapati said, they're busy with their own projects or they're still learning how to code in the **right/best way** or they're simply not interested.

**Comment by sreeram-dev** [comment]
http://stackoverflow.com/questions/8952622/why-should-i-really-move-to-coffee-script-how-much-time-does-it-really-save

**Comment by ghost** [comment]
Agree with you @sreeram-boyapati on the above thing. I got only one thing to say on this i am sure you guys are not gonna do this even while you are working right or in 3-4 years to come if you wan't others to maintain and contribute to this project why not make it more readable and easier to understand that's what coffescript was meant for in the first place

**Comment by ghost** [comment]
@aneesh-neelam and @sreeram-boyapati will fork it and create an mvp for you guys

**Comment by aneesh-neelam** [comment]
@aashrairavooru: Thank you. Please let us know how it goes. Also, we shall be documenting the proposed features, please feel free to pitch in. Until then, check out the [Issues](https://github.com/aneesh-neelam/VITacademics/issues), both open and closed and try to make improvements.

**Comment by ghost** [comment]
@aneesh-neelam will you be interested in seeing any particular files in coffescript,to see how it will be if there is i will start with them or better assign me a part of the project which i can totally shift to coffescript so that you guys can see how things are with coffee

**Comment by aneesh-neelam** [comment]
@aashrairavooru: Anything under the [/api directory](https://github.com/aneesh-neelam/VITacademics/tree/master/api) will do. Start with login, then data and then the friends feature. Also, attempt to fix #33 and other issues if possible.

**Comment by ghost** [comment]
Will try to fix the #33 my mongo is a little rusty been a while i have been using postgresql these days and i am kind of surprised you went with mongo since heroku supports postgres out of the box and even gives you the postgres dashboard in your free instance

**Comment by aneesh-neelam** [comment]
We use MongoLab, 512 MB free. Heroku PostgreSQL has a 10k row limit on the free plan.

**Comment by ghost** [comment]
10K rows aren't enough,what kind of data are you guys packing:)

**Comment by aneesh-neelam** [comment]
Check out [/api/sample-data](https://github.com/aneesh-neelam/VITacademics/tree/master/api/sample-data) and see for yourself. 

Anyway, you should know that we believe CoffeeScript is not something that is worth the time and effort. Think about it, you are talking about converting a JavaScript codebase to a CoffeeScript codebase. And to use CoffeeScript, it must be transcomplied into JavaScript. 

CoffeeScript may be less verbose and more like Python, but is it truly necessary? Its not difficult to maintain a JavaScript codebase at all, no more than a CoffeeScript one. Also the number of people who know JavaScript is much higher. You may be dettering new contributors by using a less familiar language. 

Also, you generally do not learn much if you use a language which does the hand-holding for you. There is a reason people are taught C/C++ first. 

New projects may be written using CoffeeScript, but rewriting one? Think about it, and please do not waste your time and effort. We value developer resources.

**Comment by ghost** [comment]
Coffescript can be learnt in an hour like seriously an hour without any problems and to be frank give me more javascript projects to convert to coffescript thats the only language in the world i want to code in its that beautiful and trust me show a new developer a coffescript code and a javscript code and tell me which one he understands more quickly seriously do this exercise and see for yourself

**Comment by aneesh-neelam** [comment]
Rewriting an entire codebase just to make it look beautiful seems too... I don't know. 

An hour? You generally don't learn anything in an hour. You figure out the syntax maybe, but that's about it. And that may be true only if someone knows JavaScript/Node.js pretty well in the first place. 

JavaScript specifics like callbacks, control flow, prototypes, promises, closures etc. can't be learnt in an hour. You have to make sure all JavaScript code either written or transcompiled conforms to the latest ECMAScript for best performance and compatibility. What's more, you may observe that I use array.forEach(callback) rather than a simple for loop. Its slower than a for loop, but it doesn't hog the Node.js event loop, allowing more number of operations to execute or more HTTP requests to be served asynchronously. These kinds of optimizations you cannot do on CoffeeScript. 

Might I suggest you try out [Golang](https://golang.org/) instead. Go code is compiled to native code and it would certainly increase performance. And the Go compiler is very strict, keeping the code quality high. @sreeram-boyapati and I might even help you with that.

**Comment by aneesh-neelam** [comment]
You want JavaScript projects? Start off with these: 
- [karthikb351/CaptchaParser](https://github.com/karthikb351/CaptchaParser)
- [karthikb351/AutoCaptcha-for-Chrome](https://github.com/karthikb351/AutoCaptcha-for-Chrome)
- [karthikb351/VITacademics-for-Web](https://github.com/karthikb351/VITacademics-for-Web)

VITacademics-for-Web is somewhat large and complex, you can try it out first. Its also a high priority.

**Comment by ghost** [comment]
I meant it takes a javascript developer an hour to learn coffee. Why would i shed the freedom to code everything in a single language by going with GO but thats a talk for some other time and place

**Comment by biocross** [comment]
Boy, you really are in love with coffeescript.

**Comment by ghost** [comment]
Try it tell me if you don't fall in love with it too

**Comment by biocross** [comment]
@aashrairavooru 

The fact that you can understand the javascript code to convert it to coffeescript assures me it's readable enough.

Also, we're a small group of people who work on these projects whenever we can, so we would really appreciate it if you would use your energy in implementing something new or fixing bugs, instead of re-writing what's already done. 

We have a lot of unimplemented ideas for VIT, and I'm sure you would too. We would love it if you work with us on something new.
